### PR TITLE
[Enhancement] fix status check issue in spill-related code

### LINKED ISSUE #27471: [Proposal] make sure each returned Status is handled explicitly
Status is not handled properly in many places in our code. It is easy to find such problems through error injection, such as #27033 , #27160 , and #26291.
Once this type of error occurs, it may lead to errors in subsequent data processing, and it is difficult to troubleshoot when problems occur.
So I propose to add the nodiscard keyword to Status to ensure that each returned Status can be clearly processed, otherwise a warning will be generated when compiling

https://github.com/StarRocks/starrocks/pull/27411 is the work I am doing, but there are a lot of changes, and it needs to be gradually split into multiple PRs.

### PR DESCRIPTION
**Pull request by silverbullet233**
Fix #27471

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

File: be/src/exec/spill/mem_table.h [hunk 1]
```diff
@@ -59,8 +59,8 @@ class SpillableMemTable {
     virtual bool is_empty() = 0;
     size_t mem_usage() { return _tracker->consumption(); }
     // append data to mem table
-    virtual Status append(ChunkPtr chunk) = 0;
-    virtual Status append_selective(const Chunk& src, const uint32_t* indexes, uint32_t from, uint32_t size) = 0;
+    virtual void append(ChunkPtr chunk) = 0;
```
Comment (stdpain) [inline be/src/exec/spill/mem_table.h:62]: append-related code needs support return Status code

File: be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp [hunk 2]
```diff
@@ -110,7 +110,7 @@ Status SpillableHashJoinBuildOperator::set_finishing(RuntimeState* state) {
                 state, *io_executor, TRACKER_WITH_SPILLER_GUARD(state, spiller));
     };
 
-    publish_runtime_filters(state);
+    RETURN_IF_ERROR(publish_runtime_filters(state));
```
Comment (stdpain) [inline be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp:113]: publish runtime filter error is not a critical error. we could ignore it. SpillProcessTasksBuilder shouldn't be ignored

File: be/src/exec/spill/mem_table.h (hunk 1 above)
Comment (silverbullet233) [reply be/src/exec/spill/mem_table.h:62]: > append-related code needs support return Status code

why? actually, it always return OK for now.

File: be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp (hunk 2 above)
Comment (fzhedu) [reply be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp:113]: add a warning?

File: be/src/exec/spill/spill_components.cpp [hunk 3]
```diff
@@ -184,24 +181,22 @@ Status PartitionedSpillerWriter::reset_partition(const std::vector<const SpillPa
...
-    return Status::OK();
+    _init_with_partition_nums(state, num_partitions);
+    _prepare_partitions(state);
 }
 
-Status PartitionedSpillerWriter::_init_with_partition_nums(RuntimeState* state, int num_partitions) {
+void PartitionedSpillerWriter::_init_with_partition_nums(RuntimeState* state, int num_partitions) {
     DCHECK((num_partitions & (num_partitions - 1)) == 0);
     DCHECK(num_partitions > 0);
     DCHECK(_level_to_partitions.empty());
```
Comment (fzhedu) [inline be/src/exec/spill/spill_components.cpp:202]: what if these `DCHECK` is rewriteen to `return error status`?

File: be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp (hunk 2 above)
Comment (silverbullet233) [reply be/src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp]: fixed, change to to WARN_IF_ERROR

File: be/src/exec/spill/spill_components.cpp (hunk 3 above)
Comment (silverbullet233) [reply be/src/exec/spill/spill_components.cpp:202]: > what if these `DCHECK` is rewriteen to `return error status`?

These are some basic assumptions and constraints. If they are not met, it must be an internal implementation bug. I think adding these DCHECK to verify it during test is enough, not necessary to return status.

File: be/src/exec/spill/mem_table.h (hunk 1 above)
Comment (stdpain) [reply be/src/exec/spill/mem_table.h:62]: We may later use partial ordering in the append function and may need to handle the Status return result

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 6 / 39 (15.38%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp|0|2|00.00%|[113, 167]|
|:large_blue_circle:|src/exec/pipeline/nljoin/spillable_nljoin_build_operator.cpp|0|1|00.00%|[82]|
|:large_blue_circle:|src/exec/pipeline/nljoin/nljoin_context.cpp|0|1|00.00%|[98]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/spillable_hash_join_probe_operator.cpp|0|4|00.00%|[107, 109, 110, 111]|
|:large_blue_circle:|src/exec/spill/spiller.hpp|0|1|00.00%|[119]|
|:large_blue_circle:|src/exec/pipeline/hash_partition_context.cpp|0|1|00.00%|[65]|
|:large_blue_circle:|src/exec/partition/chunks_partitioner.h|0|1|00.00%|[91]|
|:large_blue_circle:|src/exec/spill/spill_components.cpp|1|17|05.88%|[45, 136, 139, 141, 162, 184, 187, 195, 196, 199, 221, 223, 482, 487, 532, 589]|
|:large_blue_circle:|src/exec/spill/spiller.cpp|2|6|33.33%|[108, 110, 113, 114]|
|:large_blue_circle:|src/exec/spill/mem_table.cpp|2|4|50.00%|[39, 89]|
|:large_blue_circle:|src/exec/spill/spill_components.h|1|1|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:disappointed: **fail** : 4 / 34 (11.76%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/pipeline/hashjoin/spillable_hash_join_build_operator.cpp|0|2|00.00%|[113, 167]|
|:large_blue_circle:|src/exec/pipeline/nljoin/spillable_nljoin_build_operator.cpp|0|1|00.00%|[82]|
|:large_blue_circle:|src/exec/pipeline/nljoin/nljoin_context.cpp|0|1|00.00%|[98]|
|:large_blue_circle:|src/exec/pipeline/hashjoin/spillable_hash_join_probe_operator.cpp|0|4|00.00%|[107, 109, 110, 111]|
|:large_blue_circle:|src/exec/pipeline/hash_partition_context.cpp|0|1|00.00%|[65]|
|:large_blue_circle:|src/exec/partition/chunks_partitioner.h|0|1|00.00%|[91]|
|:large_blue_circle:|src/exec/spill/spill_components.cpp|1|17|05.88%|[45, 136, 139, 141, 162, 184, 187, 195, 196, 199, 221, 223, 482, 487, 532, 589]|
|:large_blue_circle:|src/exec/spill/spiller.cpp|2|6|33.33%|[108, 110, 113, 114]|
|:large_blue_circle:|src/exec/spill/spill_components.h|1|1|100.00%|[]|

**Review by stdpain — APPROVED** [review]

**Review by fzhedu — APPROVED** [review]

**Pull request merged**